### PR TITLE
Remote indexes

### PR DESCRIPTION
**Pull request by guillemcortes**
Addresses #155 , #156 and #159

### CODE REVIEW

**Comment by guillemcortes** [comment]
TODO
 The goal is to mimic mirdata remote index management since it also allows version control. The only downside is that it requires to have an "index sample" for testing (like the .wav files). Check the mirdata PR: https://github.com/mir-dataset-loaders/mirdata/pull/496
 
 - [x] `Index` class in core.py
 - [x] define indexes in all loaders as INDEXES (for reference: https://github.com/mir-dataset-loaders/mirdata/blob/master/mirdata/datasets/acousticbrainz_genre.py#L63)
 - [x] Modify `Dataset` class to match [mirdat's](https://github.com/mir-dataset-loaders/mirdata/blob/master/mirdata/core.py#L74)
     - [x] remove `custom_index_path` attribute of `Dataset` class.
     - [x] Add `version`
     - [x] Add `indexes`
 - [x] Modify general `tests`
 - [x] Create sample json indexes for all soundata indexes.
 - [x] downloader(),... add remote index
 - [x] Augment this list after reviewing [mirdata's PR 496](https://github.com/mir-dataset-loaders/mirdata/pull/496)
     - [x] Modify example.py and test_example.py from the documentation
     - [x] Soundata __init__.py
     - [x] Check #TODO comments in guillemcortes' commits 
 - [x] Validate index!
 - [x] Consider remote indexes as the only option. Remove conditions
 - [x] Fix index storage path
 - [x] DOCUMENT EVERYTHING NEW :)
     - [x] As a contributor, where to upload the index
     - [x] As a contributor, creating sample index, where to put it, how to test, etc.
     - [x] As a user, how to download and use index
 - [x] test that all `_sample` indexes version is "sample"
 - [x] Remove package data in `setup.py`
 - [x] Remove all indexes from soundata/datasets/indexes
 - [x] Do we need any extra/new test?
 - [x] Delete "working with remote indexes" from docs
 - [x] Document Sample Indexes inside "Add tests"
 - [x] Document how to create the Sample index and how to define it, example of INDEXES variable in the loader...
 - [x] Update docs/source/contributing_examples with the remote indexes

**Comment by guillemcortes** [comment]
If we follow mirdata, they store sample indexes in https://github.com/mir-dataset-loaders/mirdata/blob/master/mirdata/datasets/indexes/acousticbrainz_genre_index_sample.json `mirdata/datasets/indexes/index_sample.json` since we plan to have all indexes hosted in Zenodo, I think it makes more sense if the sample indexes that are only used for testing are stored in `tests/indexes/`. What do you think? @genisplaja @tanmayy24

File: soundata/datasets/indexes/dcase_birdVox20k_index_1.0_sample.json
Comment (genisplaja) [inline soundata/datasets/indexes/dcase_birdVox20k_index_1.0_sample.json]: `1.0` --> `sample`

**Comment by guillemcortes** [comment]
Hi @hadware, @hagenw,

We would like to invite you to review this PR that addresses the issues you raised during JOSS review: #155 #156 and #159 .

Summarizing the PR:
- Indexes will be stored at https://zenodo.org/communities/audio-data-loaders/ from now on.
- Sample indexes for testing will have to be created and packaged with Soundata (similarly to what we do with audio files for testing)
- Created an Index class that supports different versions of a dataset. (We adopted the same approach as in [mirdata](https://github.com/mir-dataset-loaders/mirdata/pull/496).
- Document the new process for both users and contributors.
- Add tests accordingly :)

Any comments/suggestions/concerns you might have would be really helpful to us.

File: docs/source/contributing.rst
Comment (tanmayy24) [inline docs/source/contributing.rst:308]: Should we say something about what is dataset-id?

File: docs/source/contributing.rst
Comment (tanmayy24) [inline docs/source/contributing.rst:295]: Screenshot 2024-05-23 at 10 25 15 AM
In this we see that even though it downloads the index in the default folder, it says it's downloading the index in the initialized folder

File: docs/source/contributing.rst
Comment (tanmayy24) [inline docs/source/contributing.rst:309]: Don't we have to tell about the license or would we edit it once we get the request?

**Comment by hadware** [comment]
Perfect! This appears to perfectly address these issues.

I would advise cleaning up the git tree after you've merged your PR (using https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/removing-sensitive-data-from-a-repository , this _will_ required a force-push at some point, as it rewrites the diff tree).

**Review by hadware — APPROVED** [review]

File: docs/source/contributing.rst
Comment (genisplaja) [reply docs/source/contributing.rst]: I think that now the message should be improved. Can you please check @tanmayy24?

Comment (tanmayy24) [reply docs/source/contributing.rst:295]: [](url)
Screenshot 2024-05-24 at 9 25 14 AM
Yes, it works fine but when we are doing download for index it shouldn't say downloading rest files at /testing/

Comment (genisplaja) [reply docs/source/contributing.rst]: indeed. Fixed now :) Maybe it's a little bit overkill to use `if` `elif` `else` but that's the easiest way I found to log more meaningful messages here.

Comment (tanmayy24) [reply docs/source/contributing.rst:295]: Thank you! Looks good to me :)))

**Comment by guillemcortes** [comment]
This looks good! Should we merge? @genisplaja @tanmayy24

File: docs/source/contributing.rst
Comment (hagenw) [inline docs/source/contributing.rst]: This sentence seems to make sense only for already existing indices. I guess new ones are never added to the repo in the first place:

```suggestion
We store all indexes in an online repository on Zenodo.
```

File: docs/source/tutorial.rst
Comment (hagenw) [inline docs/source/tutorial.rst:59]: When is no canonical version of the dataset available?
When there are more than one version?
What happens then, will the latest version be downloaded automatically, or is the user forced to specify a version?

**Comment by hagenw** [comment]
Nice to see, that this also helps addressing the problem of dataset versioning.

If the merge request should be ready for merging, I would recommend to remove `[WIP]` from its title ;)

File: docs/source/contributing.rst
Comment (genisplaja) [reply docs/source/contributing.rst]: hey @hagenw, done! Thanks for the suggestion.

File: docs/source/tutorial.rst
Comment (genisplaja) [reply docs/source/tutorial.rst:59]: Ok, thanks! Added documentation and examples about these issues. To better support the use of different versions, I did add a `list_dataset_versions()` function in `soundata.__init__` which should help users get the available versions before initializing the data loaders. This is documented as well. This is all in my last commit.

Does that address the issue @hagenw? Please @tanmayy24 and @guillemcortes also take a look to see if that sounds good to you.

File: docs/source/tutorial.rst
Comment (hagenw) [inline docs/source/tutorial.rst]:
```suggestion
Some datasets have more than one version available. Soundata can handle more than one version if included in the dataloader.
```

BTW, I would highly recommend to use semantic linefeeds in the future, when writing documentation (sorry, can only share as an image):

image

File: docs/source/tutorial.rst
Comment (hagenw) [inline docs/source/tutorial.rst]:
```suggestion
You may run ``soundata.list_dataset_versions('urbansound8k')`` to see all available versions of the dataset, that are available through its dataloder.
```

File: docs/source/tutorial.rst
Comment (hagenw) [inline docs/source/tutorial.rst]: Latest refers to the latest published version, not the on with the highest version number, correct?
E.g. assume you have a dataset with versions 1.0 and 2.0, and then you publish a fix for 1.0 as 1.1, then I guess `latest` refers to 1.1.

File: docs/source/tutorial.rst
Comment (hagenw) [reply docs/source/tutorial.rst:59]: Thanks, for updating the docs accordingly.
Looks fine to me, I just added a few small comments.

File: docs/source/tutorial.rst
Comment (genisplaja) [reply docs/source/tutorial.rst]: That's right @hagenw. I was trying to mention that we point to the _highest_ version. Therefore, in the example you posed, the default would still point to 2.0. I actually wonder if it is actually right to assume that we should be pointing to the highest version, probably not. It will most likely depend on each case. 

One thing that we could do, is letting the data loader creator to decide which version should be defaulted, and then update the `list_dataset_versions()` so that instead of returning the list of version numbers, displays a message like:

```
Dataset <dataset-name> has the following versions available: ["1.0", "1.1", "2.0"].
Default version for <dataset-name> is "2.0".
```
Does that sound good? Of course we would document these change as such.

Comment (genisplaja) [reply docs/source/tutorial.rst]: Nonetheless, if we want a criteria for that issue we could ask the contributors to default to the highest version unless there is a motivation to point to a different one (maybe the highest one is really recent and most of the works are based on the old one, etc.)

File: docs/source/tutorial.rst
Comment (genisplaja) [reply docs/source/tutorial.rst]: Ok! I am waiting until we resolve the versioning issue here: https://github.com/soundata/soundata/pull/169#discussion_r1616019129, since it'll most likely make this line change again :)

File: docs/source/tutorial.rst
Comment (genisplaja) [reply docs/source/tutorial.rst]: I see. Indeed, thanks for the suggestion. Actually, longer lines make editing more difficult as well, since you cannot read the entire thing and you have to move left and right a lot. Will take into account for sure!

File: docs/source/tutorial.rst
Comment (hagenw) [reply docs/source/tutorial.rst]: I think allowing to set a default version by the dataset maintainer makes sense, otherwise you cannot choose a default with a lower version number.

File: docs/source/contributing.rst
Comment (tanmayy24) [inline docs/source/contributing.rst:1]: pip install .
        pip install ."[tests]"
        pip install ."[docs]"
        
I think we should separately mention the use case for each. Like it's a common tendency to install all of them but rather separating them out for each use case should help
e.g

Install Core Dependencies:
pip install .
Install testing dependencies:
pip install ."[tests]"
Install Documentation Dependencies:
pip install ."[docs]"

**Review by tanmayy24 — APPROVED** [review]
Excellent work on this fix! Appreciate your effort! :))

**Comment by faroit** [comment]
wow, this is cool

**Pull request merged**
